be/src/exec/pipeline/exchange/mem_limited_chunk_queue.cpp

```diff
@@ -0,0 +1,528 @@
+    _opened_source_opcount[consumer_index] -= 1;
+    if (_opened_source_opcount[consumer_index] == 0) {
+        _opened_source_number--;
+        _close_consumer(consumer_index);
+    }
+}
+
+void MemLimitedChunkQueue::open_producer() {
+    std::unique_lock l(_mutex);
+    _opened_sink_number++;
+}
+
+void MemLimitedChunkQueue::close_producer() {
+    std::unique_lock l(_mutex);
+    _opened_sink_number--;
+}
+
+void MemLimitedChunkQueue::_update_progress(Iterator* iter) {
+    if (iter != nullptr) {
+        Cell* cell = iter->get_cell();
+        _fastest_accumulated_rows = std::max(_fastest_accumulated_rows, cell->accumulated_rows);
+        _fastest_accumulated_bytes = std::max(_fastest_accumulated_bytes, cell->accumulated_bytes);
+    } else {
+        // nullptr means one consumer is closed, should update
+        _fastest_accumulated_rows = 0;
+        _fastest_accumulated_bytes = 0;
+        for (int32_t i = 0; i < _consumer_number; i++) {
+            auto iter = _consumer_progress[i].get();
+            if (iter == nullptr) {
+                continue;
+            }
+            if (!iter->has_next()) {
+                // all data is consumed
+                _fastest_accumulated_rows = _total_accumulated_rows;
+                _fastest_accumulated_bytes = _total_accumulated_bytes;
+                break;
+            }
+            Cell* cell = iter->get_cell();
+            _fastest_accumulated_rows = std::max(_fastest_accumulated_rows, cell->accumulated_rows);
+            _fastest_accumulated_bytes = std::max(_fastest_accumulated_bytes, cell->accumulated_bytes);
+        }
+    }
+
+    while (_iterator.valid()) {
+        Cell* cell = _iterator.get_cell();
+        if (cell->used_count != _consumer_number) {
+            break;
+        }
+        _head_accumulated_rows = cell->accumulated_rows;
+        _head_accumulated_bytes = cell->accumulated_bytes;
+        // advance flushed position
+        _flushed_accumulated_rows = std::max(_flushed_accumulated_rows, _head_accumulated_rows);
+        _flushed_accumulated_bytes = std::max(_flushed_accumulated_bytes, _head_accumulated_bytes);
+        VLOG_ROW << fmt::format("release chunk, current head_accumulated_rows[{}], head_accumulated_bytes[{}]",
+                                _head_accumulated_rows, _head_accumulated_bytes);
+        cell->chunk.reset();
+        if (!_iterator.has_next()) {
+            break;
+        }
+        _iterator.move_to_next();
+    }
+}
+
+void MemLimitedChunkQueue::_close_consumer(int32_t consumer_index) {
+    DCHECK(_consumer_progress[consumer_index] != nullptr);
+    auto iter = _consumer_progress[consumer_index].get();
+    if (iter->has_next()) {
+        do {
+            iter->move_to_next();
+            Cell* cell = iter->get_cell();
+            cell->used_count += 1;
+        } while (iter->has_next());
+    }
+    _consumer_progress[consumer_index].reset();
+    _update_progress();
+}
+
+Status MemLimitedChunkQueue::_flush() {
```

Comment:
we prefer make this function support yield

Reply:
> we prefer make this function support yield

yield may be not necessary in this operator because block is already small enough. If necessary, we can support it in the future